### DISCUSSION TITLE
Feedback on app.config

### DISCUSSION THREAD

**Discussion by Tahul** [nuxt/nuxt#17045] [General]
Hey,

I tried app.config feature that has been merged in framework, from a basic project and this branch in Docus that replaces @nuxt-themes/config by the usage of appConfig ; here are my few feedbacks on it.

Some references:
- [@nuxt-themes/config](https://github.com/nuxt-themes/config)
- [Docus PR that replaces @nuxt-themes/config by appConfig](https://github.com/nuxt-themes/docus/pull/614/commits/5c64def9ce214a116cfae33cb157168108d93c3b#diff-da36bd3b1d80b8f350201bef8bfce626fe1730550381c8dfeff8be6bcb9a91ca)

---

## 1. API

I've seen the PR mentions a crossed task:

```
Inject $appConfig (A plugin can easily inject reactive shortcut. Delegating addition to next steps)
```

I think this is one of the reason to exist of this feature, I would love to make it happen with some context.

[@nuxt-themes/config](https://github.com/nuxt-themes/config) existed to strictly enforce a way to declare and access the configuration for a theme.

Enforcing it makes it easier for an editor to be build on top of this feature, not only an editor for the actual config file, but also an editor that allows the injection of the available keys from this config file into specific slots such as components props.

Also, having an enforced way to access these variables makes it really easy for us to detect the presence of these variables inside any `.vue` file in the user app, which makes it easy for modules like [nuxt-component-meta](https://github.com/nuxtlabs/nuxt-component-meta) to enrich the Studio experience based on these scoped datas.

With current version of appConfig, we can expect theme authors to implement these logics themselves (as suggested, via a $appConfig shortcut, or [any other non-standard way to do it](https://github.com/nuxt-themes/docus/pull/614/commits/5c64def9ce214a116cfae33cb157168108d93c3b#diff-2e1548347de388abff6ff23e4a3f4b3d91dd09883da1dc9f292935e40248ea4aR1-R3)) which removes most of the purposes of the work done on previous iterations.

What we had in `@nuxt-themes/config` was `$theme('my.config.key.path')` available everywhere, with a proper type generated for `my.config.key.path` for every available key in the theme configuration object.

Not only this was super convenient to use in any place of the app, but also was super easy to parse in the component code AND imagine being a kind of input inside a Studio form.

## 2. Deep merging strategy

As seen [here](https://github.com/nuxt/framework/pull/6333/files#diff-74e5c3bbf838d41871a150093f241165be7796a00f5fc6ad4976b6d8bc5ea27eR119-R125), the `app.config` file will follow the same merging strategy as `extends` feature, which is not exactly what we want for this config specifically.

This config is made to be partially overwritten by the end-user, while fully declared by the theme authors.

One great example is icons, that you can see here:
- [nuxt-themes/docus@`5c64def` (#614)](https://github.com/nuxt-themes/docus/pull/614/commits/5c64def9ce214a116cfae33cb157168108d93c3b#diff-360243eeb21c7364e0bb2314d366f247bf11f96e7e652cd948b60aa14025a4f9R28)
- [nuxt-themes/docus@`5c64def` (#614)](https://github.com/nuxt-themes/docus/pull/614/commits/5c64def9ce214a116cfae33cb157168108d93c3b#diff-da36bd3b1d80b8f350201bef8bfce626fe1730550381c8dfeff8be6bcb9a91caR6-R19)

In that example, you can see that the theme.config file does not allow setting defaults for any key that is an array, as `defu` default behavior will lead to merging these arrays, which results in having both the defaults defined by the theme author and the values written by the theme user.

Maybe it's okay but I think it should be clearly stated that theme authors cannot define defaults for array keys or they will be shipped to the theme user project anyways.

## 3. Schema support

`theme.config` file has support for `export const schema` that would allow theme authors to do deeper in the definition of their theme configuration.

This schema followed [untyped](https://github.com/unjs/untyped) schema for now, but was thought to be improved in order to support form generation feature from Studio.

Not only this allows to improve the typing and general developer experience for theme authors in the long run, but this also was built with the Studio in mind.

Exposing a schema makes it a lot easier for us to build an editor on top of this configuration.

A configuration key in VSCode is extremely simple to edit, but building a whole human-readable form around it gives the need to encapsulate it around other datas.

I'm afraid that without being able to specify a schema for the theme configuration, we won't be able to make the form look great for end-users (=== users of the Studio).

Also, schema makes it a lot easier to recompose a valid JS/TS file for configuration from a form, which also makes it easier for us to skip having to maintain AST logics for this to happen.

## 4. Metas

This is a side-purpose, but I think we should start thinking on how people should define values like `title`, `author`, and other meta-values for their theme, being able to watermark a theme.

Dropping [@nuxt-themes/config](https://github.com/nuxt-themes/config) in favor of this feature also completely removes this concept from the stack, which has I think some value.

Theme authors can watermark their theme.

Theme users can know what theme they are using when the project boot.

This was built with the intention of not making it look like an extended Nuxt project was "vanilla" Nuxt, as it can easily be mistaken if this information is not provided somewhere else.

But also, this also was a way to enforce some ways to supply some data, as the Studio in the long run should I think be able to provide some kind of marketplace around that theming feature, which I think cannot be properly built without providing these datas in a standard way from the theme itself.

Feeding both the direct output for developers and the end-user UI with a single source of truth for these datas seemed the good approach.

## 5. Extendability

I see that documentation for the feature says that we can define a `theme` key inside app.config, yet I don't really understand what is the point of other keys in that config file?

Do we want module authors to actually use that configuration file instead of a `nuxt.config` key?

What is the proper way to reserve a key in that config file?

In [these lines](https://github.com/nuxt/framework/pull/6333/files#diff-74e5c3bbf838d41871a150093f241165be7796a00f5fc6ad4976b6d8bc5ea27eR118-R125) I can see that I could somehow extends these config files, how should we do that exactly?

---

This feature sounds like a lower-level / more generic approach than what `@nuxt-themes/config` is, but not being generic was most of the purpose of that module.

If we want to ship some valuable (marketable) features to a user-base with a product, we ultimately have to narrow down the scopes of how people should use our tools that are built to be generic, or it will take years to cover all the possible use-cases with a product that provides an understandable user-experience.

What I expected from integrating the concepts of reactive/hmr-compatible configuration files into Nuxt core was to have a set of utils that would have make my life easier when it comes to add a configuration file linked to the Nuxt project.

I thought it was going to kill the need for modules such as Tailwind, Design Tokens, nuxt-themes/config, and probably others, to maintain their own way of loading configuration files, and that sounded amazing.

I would love to get guidelines on how to remove or adapt `@nuxt-themes/config` to that new features, while preserving the features.

Thanks for reading the issue and for all the efforts put into pushing these features into the framework! 😄

**Comment by atinux** [comment]
Thank for you this early feedback @Tahul ❤️ 

## 1.API

I believe having a `$appConfig('my.config.key.path')` is possible, as well as `useAppConfig('my.config.key.path')` (also related to #6776 ).

I don't see directly the link to Studio in the UI though until we give the possibility to inject app config as variables.

## 2. Deep merging strategy

Indeed, we may want to use `defu.arrayFn` for this to avoid merging those array, WDYT @pi0 ?

## 3. Schema support

I agree that we shall support `untyped` as schema for giving better typing and also generating a form (like we do for the `nuxt.config.ts`), how do you see a possible implementation for it @pi0 @danielroe?

## 4. Metas

I believe this is the responsibility of Nuxt and the `extends` feature, the metas are directly part of the `package.json` (name, description and version) and is widely used as a standard. Shall we add an issue to support this?

## 5. Extendability

> I see that documentation for the feature says that we can define a theme key inside app.config, yet I don't really understand what is the point of other keys in that config file?

By working on Docus, I am now thinking that having multiple keys is even better. Docus will directly extends from Nuxt Elements. Then how to configure both of them using a `theme` key?

Instead, I believe that Docus could leverage the `docus` key and Nuxt Elements the `elements` key to be configured directly. The beauty of the extends is that it can leverage multiple projects 🚀 

```ts
// app.config.ts
export default defineAppConfig({
  docus: {
    title: 'Nuxt Content',
    socials: {
      twitter: '@nuxt_js',
      github: 'nuxt/content'
    },
  elements: {
    // elements config
  }
})
```

This allows mixing multiple themes.

**Comment by pi0** [comment]
I really appreciate your time on trying and writing feedback @Tahul. This definitely helps to improve the core API.

Regarding allowing to access a scope, https://github.com/nuxt/framework/pull/6776 tries to implement it for one level at least. I believe encouraging to destructure more than 1 levels, is probably not the best idea since it will make it harder to reason about source of config and lead to less explicit code. Also runtime addition to core to parse dot notion.

For merging, good point we could use `defu.arrayFn` (or was thinking even a custom merger with special `_merge` hook for module authors)

Regarding schema generation, I fully agree with you and thinking to introduce an `schema` key from `nuxt.config` that allows generating schema by declaration. It can _also_ be used by module authors to define config schema and project such as Nuxt studio and devtools leveraging it for nice editor UI. It cannot be defined from `app.config` (well at least without hacking to tree-shake) because this file is runtime bundled. (I will probably push-in another PR. we completely switched to new `Defu` helper but removed untyped in last commits)

Regarding meta's, of course, you are free to use `.theme` namespace but also I would suggest (same as modules) themes specify such meta into their keys like `.docus` to avoid conflicting. Generic meta such as `title` are subject to be in standard place of core types so that we could reuse them for several purposes such as SEO meta and PWA manifest. And rest of meta, belong to _nuxt config_ layer for extends.

And the rest as @Atinux mentioned (we were writing same time)

**Reply by atinux** [reply]
>  It cannot be defined from app.config (well at least without hacking to tree-shake) because this file is runtime bundled. (I will probably push-in another PR. we completely switched to new Defu helper but removed untyped in last commits)

Could is be possible to define the schema using the `$` prefix?

```ts
export default defineAppConfig({
  docus: {
    asideLevel: 1
  },
  $docus: {
    asideLevel: {
      $resolve: val => Number(val),
      $default: 1
    }
  }
})
```

Then I guess regarding tree-shaking, we can remove every key starting with `$` before adding it to the bundle?

**Reply by pi0** [reply]
We are actually directly bundling them. For supporting such syntax, we need a ast parser plugin that can tree-shake but it makes it slower while writing schema probably is not common enough to need HMR and can be from nuxt.config

**Comment by Tahul** [comment]
@Atinux answers;

> I don't see directly the link to Studio in the UI though until we give the possibility to inject app config as variables.

Not sure about sharing screenshots of Studio UI, but I imagine for component props we could've had autocompleting inputs that feeds from project datas.

Thinking of this kind of UI:
![CleanShot 2022-08-19 at 12 22 43@2x](https://user-images.githubusercontent.com/1385263/185598925-8fb713b9-3cd2-4989-ad9d-0a2618c43db0.png)

This is possible if we use this kind of syntax, just by exposing the list of possible paths (and most likely filled values) via an API.

That is obviously some first-step integration for the inline variable feature @farnabaz built in this PR:
https://github.com/nuxt/content/pull/1266

> I believe this is the responsibility of Nuxt and the extends feature, the metas are directly part of the package.json (name, description and version) and is widely used as a standard. Shall we add an issue to support this?

If we limit our marketplace data to be strictly what NPM does offer as well, then I think we could drop it and base boot output from extends feature gathering package.json datas.

Yet this is not what we do with [modules](https://github.com/nuxt/modules/tree/main/modules) and I think themes are even more "marketable" and "visual" than modules, in the sense they are more UI-focused.

@pi0 answers;

> Also runtime addition to core to parse dot notion.

This is all the runtime addition needed for it to happen:

```
const get = (obj, path, defValue = undefined) => {
  if (!path) return undefined
  const pathArray = Array.isArray(path) ? path : path.match(/([^[.\\]])+/g)
  const result = pathArray.reduce(
    (prevObj, key) => prevObj && prevObj[key],
    obj
  )
  return result === undefined ? defValue : result
}
```

> It cannot be defined from app.config (well at least without hacking to tree-shake) because this file is runtime bundled.

Then it will be hard to write & maintain, if it is fully based on nuxt.config file, it is currently pretty hard and inconvenient to work on the file for big sessions, as most of the time it needs full server restart for the changes to be applied.

I imagine the experience of writing my config schema by having the studio or devtools running locally, and seeing the actual form the user of my theme will be using being reloaded via HMR on each of my changes on my schema.

> Regarding meta's, of course, you are free to use .theme namespace but also I would suggest (same as modules) themes specify such meta into their keys like .docus to avoid conflicting. Generic meta such as title are subject to be in standard place of core types so that we could reuse them for several purposes such as SEO meta and PWA manifest. And rest of meta, belong to nuxt config layer for extends.

I think there is a quiproquo here 😄, what I'm describing as "metas" is theme description data, that are defined by the **author** of the theme and not overwriteable by the end user.

Can be seen for Docus [here](https://github.com/nuxt-themes/docus/blob/dev/nuxt.config.ts#L65-L69).

Data is currently limited, but might be improved in long-run to feed marketplace of datas as suggested in previous lines of this comment.

**Reply by pi0** [reply]
- The kind of UI you could imagine for studio and devtools is what untyped is made for. Obviously not limited to app config I want all config to be fully editable visually :)
- Layer meta could be in `nuxt.config` for themes. It is one-time thing and not required in runtime for things like theme author. There are other use-cases for meta to be in config layer you can discover later from c12 ;) 
- You obviously don't have to show me the code snippet how get-prop can be implemented (!) overhead is **running function** but the main reason we support one level is to avoid losing the source of trust by destructing config. You can implement any custom logic you like inside themes for custom patterns (while I discourage it and you should be aware of reactivity issues it might cause)

Declaimer for public readers: Non of framework roadmap decisions are made for a private use-case but generic possible APIs.

**Reply by atinux** [reply]
> If we limit our marketplace data to be strictly what NPM does offer as well, then I think we could drop it and base boot output from extends feature gathering package.json datas.

Well I believe we will have `nuxt/themes` repository similar to `nuxt/modules` that can fetch data from `package.json` like it does with the ability to give screenshots and other metadata. Reading from the `nuxt.config.ts` was anyway harder to the themes listing to be done easily.
